### PR TITLE
[SPARK-27393][SQL] Show ReusedSubquery in the plan when the subquery is reused

### PR DESCRIPTION
**Pull request by gatorsmile**
## What changes were proposed in this pull request?
With this change, we can easily identify the plan difference when subquery is reused. 

When the reuse is enabled, the plan looks like 
```
== Physical Plan ==
CollectLimit 1
+- *(1) Project [(Subquery subquery240 + ReusedSubquery Subquery subquery240) AS (scalarsubquery() + scalarsubquery())#253]
   :  :- Subquery subquery240
   :  :  +- *(2) HashAggregate(keys=[], functions=[avg(cast(key#13 as bigint))], output=[avg(key)#250])
   :  :     +- Exchange SinglePartition
   :  :        +- *(1) HashAggregate(keys=[], functions=[partial_avg(cast(key#13 as bigint))], output=[sum#256, count#257L])
   :  :           +- *(1) SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData, true])).key AS key#13]
   :  :              +- Scan[obj#12]
   :  +- ReusedSubquery Subquery subquery240
   +- *(1) SerializeFromObject
      +- Scan[obj#12]
```

When the reuse is disabled, the plan looks like 
```
== Physical Plan ==
CollectLimit 1
+- *(1) Project [(Subquery subquery286 + Subquery subquery287) AS (scalarsubquery() + scalarsubquery())#299]
   :  :- Subquery subquery286
   :  :  +- *(2) HashAggregate(keys=[], functions=[avg(cast(key#13 as bigint))], output=[avg(key)#296])
   :  :     +- Exchange SinglePartition
   :  :        +- *(1) HashAggregate(keys=[], functions=[partial_avg(cast(key#13 as bigint))], output=[sum#302, count#303L])
   :  :           +- *(1) SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData, true])).key AS key#13]
   :  :              +- Scan[obj#12]
   :  +- Subquery subquery287
   :     +- *(2) HashAggregate(keys=[], functions=[avg(cast(key#13 as bigint))], output=[avg(key)#298])
   :        +- Exchange SinglePartition
   :           +- *(1) HashAggregate(keys=[], functions=[partial_avg(cast(key#13 as bigint))], output=[sum#306, count#307L])
   :              +- *(1) SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData, true])).key AS key#13]
   :                 +- Scan[obj#12]
   +- *(1) SerializeFromObject
      +- Scan[obj#12]
```

## How was this patch tested?
Modified the existing test.

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @cloud-fan @maryannxue @hvanhovell

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -113,33 +113,6 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
     }
   }
 
-  test("Reuse Subquery") {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: moved to SubquerySuite.scala

**Comment by adrian-wang** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala [hunk 2]
```diff
@@ -58,11 +61,11 @@ abstract class SubqueryExpression(
         children.zip(p.children).forall(p => p._1.semanticEquals(p._2))
     case _ => false
   }
-  def canonicalize(attrs: AttributeSeq): SubqueryExpression = {
+  override def canonicalize(attrs: AttributeSeq): SubqueryExpression = {
     // Normalize the outer references in the subquery plan.
     val normalizedPlan = plan.transformAllExpressions {
       case OuterReference(r) => OuterReference(QueryPlan.normalizeExprId(r, attrs))
-    }
+    }.canonicalized
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala]: We need to the canonicalized plan here? It seems all the canonicalization should happen in line 69?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 3]
```diff
@@ -25,7 +25,7 @@ import java.util.concurrent.atomic.AtomicBoolean
 import org.apache.spark.{AccumulatorSuite, SparkException}
 import org.apache.spark.scheduler.{SparkListener, SparkListenerJobStart}
 import org.apache.spark.sql.catalyst.util.StringUtils
-import org.apache.spark.sql.execution.{aggregate, ScalarSubquery, SubqueryExec}
+import org.apache.spark.sql.execution.aggregate
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: nit: we can remove this import by changing `aggregate.HashAggregateExec` -> `HashAggregateExec` in line 263?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala [hunk 4]
```diff
@@ -715,3 +726,29 @@ object SubqueryExec {
...
+    child.prepare()
+  }
+
+  protected override def doExecute(): RDD[InternalRow] = {
+    child.execute()
+  }
+
+  override def executeCollect(): Array[InternalRow] = {
+    child.executeCollect()
+  }
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: super nit: 
```
  protected override def doPrepare(): Unit = child.prepare()
  protected override def doExecute(): RDD[InternalRow] = child.execute()
  override def executeCollect(): Array[InternalRow] = child.executeCollect()
```

**Comment by maropu** [comment]
btw, is it not worth filing a new jira for this refactoring?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala [hunk 5]
```diff
@@ -660,21 +660,28 @@ object CoalesceExec {
...
- * Physical plan for a subquery.
+ * Parent class for different types of subquery plans
  */
-case class SubqueryExec(name: String, child: SparkPlan) extends UnaryExecNode {
-
-  override lazy val metrics = Map(
-    "dataSize" -> SQLMetrics.createSizeMetric(sparkContext, "data size"),
-    "collectTime" -> SQLMetrics.createTimingMetric(sparkContext, "time to collect"))
+abstract class BaseSubqueryExec extends SparkPlan {
+  def name: String
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: We need `name` in this base class for subquery exec?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala [hunk 6]
```diff
@@ -129,13 +134,14 @@ case class ReuseSubquery(conf: SQLConf) extends Rule[SparkPlan] {
       return plan
     }
     // Build a hash map using schema of subqueries to avoid O(N*N) sameResult calls.
-    val subqueries = mutable.HashMap[StructType, ArrayBuffer[SubqueryExec]]()
+    val subqueries = mutable.HashMap[StructType, ArrayBuffer[BaseSubqueryExec]]()
     plan transformAllExpressions {
       case sub: ExecSubqueryExpression =>
-        val sameSchema = subqueries.getOrElseUpdate(sub.plan.schema, ArrayBuffer[SubqueryExec]())
+        val sameSchema =
+          subqueries.getOrElseUpdate(sub.plan.schema, ArrayBuffer[BaseSubqueryExec]())
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala]: unnecessary change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala]: +1

**Comment by cloud-fan** [comment]
Shall we create a new JIRA? This improves the SQL web UI

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 5 above)
Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: both ReusedSubqueryExec and SubqueryExec have the name

File: sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala (hunk 6 above)
Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala]: change it to BaseSubqueryExec

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
retest this please

**Comment by gatorsmile** [comment]
Thanks! Merged to master.
